nitransforms/surface.py

```diff
@@ -0,0 +1,531 @@
+                and (other.interpolation_method == self.interpolation_method)):
+            return self.__class__(
+                self.reference,
+                other.moving,
+                interpolation_method=self.interpolation_method
+            )
+        else:
+            raise NotImplementedError
+
+    def __invert__(self):
+        return self.__class__(
+            self.moving,
+            self.reference,
+            interpolation_method=self.interpolation_method
+        )
+
+    @SurfaceTransformBase.reference.setter
+    def reference(self, surface):
+        raise ValueError("Don't modify the reference of an existing resampling."
+                         "Create a new one instead.")
+
+    @SurfaceTransformBase.moving.setter
+    def moving(self, surface):
+        raise ValueError("Don't modify the moving of an existing resampling."
+                         "Create a new one instead.")
+
+    def apply(self, x, inverse=False, normalize="element"):
+        """Apply the transform to surface data.
+
+        Parameters
+        ----------
+        x : array-like, shape (..., nv1), or SurfaceMesh
+            Data to transform or SurfaceMesh to resample
+        inverse : bool, default=False
+            Whether to apply the inverse transform. If True, ``x`` has shape
+            (..., nv2), and the output will have shape (..., nv1).
+        normalize : {"element", "sum", "none"}, default="element"
+            Normalization strategy. If "element", the scale of each value in
+            the output is comparable to each value of the input. If "sum", the
+            sum of the output is comparable to the sum of the input. If
+            "none", no normalization is applied.
+
+        Returns
+        -------
+        y : array-like, shape (..., nv2)
+            Transformed data.
+        """
+        if normalize not in ("element", "sum", "none"):
+            raise ValueError("Invalid normalization strategy.")
+
+        mat = self.mat.T if inverse else self.mat
+
+        if normalize == "element":
+            sum_ = mat.sum(axis=0)
+            scale = np.zeros_like(sum_)
+            mask = sum_ != 0
+            scale[mask] = 1.0 / sum_[mask]
+            mat = mat @ sparse.diags(scale)
+        elif normalize == "sum":
+            sum_ = mat.sum(axis=1)
+            scale = np.zeros_like(sum_)
+            mask = sum_ != 0
+            scale[mask] = 1.0 / sum_[mask]
+            mat = sparse.diags(scale) @ mat
+
+        if isinstance(x, SurfaceMesh) or isinstance(x, str) or isinstance(x, pathlib.PurePath):
+            x = SurfaceMesh(x)
+            if not x.check_sphere():
+                raise ValueError("If x is a surface, it should be a sphere.")
+            x.set_radius()
+            rs_x = x._coords[:, 0] @ mat
+            rs_y = x._coords[:, 1] @ mat
+            rs_z = x._coords[:, 2] @ mat
+            y = SurfaceMesh.from_arrays(np.vstack([rs_x, rs_y, rs_z]).T, self.reference._triangles)
```

Comment:
I'm confused why this is decomposed. Is it a mathematical requirement, or a hack to get things working, or...?